### PR TITLE
Download studio videos for offline usage

### PR DESCRIPTION
**Pull request by vargaat**
### How the download works?
- After all courses finished their synchronizations and we have the offline available html files on disk we start the studio sync.
- We scan all html files and collect all iframes where studio is embedded.
- We authenticate for studio API:
  - We fetch LTIs for the account and search for Studio.
  - Launch Studio LTI in a headless webview.
  - Extract the user id and token with javascript.
  - Create an API entity for studio API access.
- The next step is to fetch studio video metadata for all courses.
- After this we know which media IDs we want to sync so we can delete no longer needed videos to free up disk space before downloading new videos.
- We download the actual video files, save captions and generate a preview image based on the first frame of the video. We make a basic check based on the file's size if an already downloaded video needs to be re-downloaded or not.
- After we have the videos we can replace iframes inside htmls with video tags pointing to the local mp4 files. The offline base urls of webviews were modified since studio videos are in a different folder compared to the html files.

### Limitations
- Sync happens after all course synchronizations complete so the studio video file downloads are not represented in the progress. If the last course has studio videos the progress will stay on 100% until all videos are downloaded.
- In case something goes wrong with the studio sync it won't fail the offline sync as a whole, errors from this step are not visible for the user. I added local error logs and also send errors to crashlytics for debugging purposes.
- Studio API authentication happens even if there are no Studio videos to download. This could be optimized since this task could take up a noticeable time.

### Notes
- The authentication method is a temporary one until we receive a dedicated API for it around Q4.
- I've tried to extract as many generic re-usable extensions as possible.
- Studio download sub-tasks are well separated inside a Studio folder. I'm not sure if these should be in the `CourseSync` folder or not.

refs: MBL-17707
affects: Student
release note: Studio videos are now downloaded during an offline sync.

test plan:
- Embed a studio video with subtitles inside a page, assignment and quiz.
- Sync the course for offline availability.
- Go offline and enter the page, assignment and quiz.
- Check if studio videos can be played with subtitles.

## Screenshots

<table>
<tr><th>Before</th><th>After</th></tr>
<tr>
<td></td>
<td></td>
</tr>
</table>

## Checklist

- [ ] Follow-up e2e test ticket created
- [ ] A11y checked
- [x] Tested on phone
- [ ] Tested on tablet
- [ ] Tested in dark mode
- [x] Tested in light mode
- [ ] Approve from product

### CODE REVIEW

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Build failed, skipping coverage check</td>
    </tr>
  </tbody>
</table>



<details><summary>:x: XCTest failed: CoreTests/CourseSyncCleanupInteractorTests/testApplicationOfflineDirectoryForUserDeleted</summary>


    XCTAssertFalse failed
</details>
<details><summary>:x: XCTest failed: CoreTests/CourseSyncFilesInteractorLiveTests/testFileDownloadError</summary>


    failed: caught error: "Error Domain=NSCocoaErrorDomain Code=4 "“CoreTestCase” couldn’t be removed." UserInfo={NSUserStringVariant=(
        Remove
    ), NSFilePath=/Users/vagrant/Library/Developer/CoreSimulator/Devices/46C95008-E0C7-49E6-A4C1-8EFF7538A1C1/data/Containers/Data/Application/9969B8B3-097E-4E9A-8AD9-D28E45ED202A/tmp/CoreTestCase, NSUnderlyingError=0x60001d64dad0 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}"
</details>
<details><summary>:x: XCTest failed: CoreTests/CourseSyncFilesInteractorLiveTests/testFileDownloadProgress</summary>


    failed: caught error: "Error Domain=NSCocoaErrorDomain Code=4 "“CoreTestCase” couldn’t be removed." UserInfo={NSUserStringVariant=(
        Remove
    ), NSFilePath=/Users/vagrant/Library/Developer/CoreSimulator/Devices/46C95008-E0C7-49E6-A4C1-8EFF7538A1C1/data/Containers/Data/Application/9969B8B3-097E-4E9A-8AD9-D28E45ED202A/tmp/CoreTestCase, NSUnderlyingError=0x60001d6e6d90 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}"
</details>
<details><summary>:x: XCTest failed: CoreTests/CourseSyncFilesInteractorLiveTests/testGetFiles</summary>


    failed: caught error: "Error Domain=NSCocoaErrorDomain Code=4 "“CoreTestCase” couldn’t be removed." UserInfo={NSUserStringVariant=(
        Remove
    ), NSFilePath=/Users/vagrant/Library/Developer/CoreSimulator/Devices/46C95008-E0C7-49E6-A4C1-8EFF7538A1C1/data/Containers/Data/Application/9969B8B3-097E-4E9A-8AD9-D28E45ED202A/tmp/CoreTestCase, NSUnderlyingError=0x600020c15530 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}"
</details>
<details><summary>:x: XCTest failed: CoreTests/CourseSyncFilesInteractorLiveTests/testMissingSessionError</summary>


    failed: caught error: "Error Domain=NSCocoaErrorDomain Code=4 "“CoreTestCase” couldn’t be removed." UserInfo={NSUserStringVariant=(
        Remove
    ), NSFilePath=/Users/vagrant/Library/Developer/CoreSimulator/Devices/46C95008-E0C7-49E6-A4C1-8EFF7538A1C1/data/Containers/Data/Application/9969B8B3-097E-4E9A-8AD9-D28E45ED202A/tmp/CoreTestCase, NSUnderlyingError=0x60001d6dcd50 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}"
</details>
<details><summary>:x: XCTest failed: CoreTests/CourseSyncFilesInteractorLiveTests/testNotUpToDateFile</summary>


    failed: caught error: "Error Domain=NSCocoaErrorDomain Code=4 "“CoreTestCase” couldn’t be removed." UserInfo={NSUserStringVariant=(
        Remove
    ), NSFilePath=/Users/vagrant/Library/Developer/CoreSimulator/Devices/46C95008-E0C7-49E6-A4C1-8EFF7538A1C1/data/Containers/Data/Application/9969B8B3-097E-4E9A-8AD9-D28E45ED202A/tmp/CoreTestCase, NSUnderlyingError=0x600020c811d0 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}"
</details>
<details><summary>:x: XCTest failed: CoreTests/CourseSyncFilesInteractorLiveTests/testUnavailableFiles</summary>


    failed: caught error: "Error Domain=NSCocoaErrorDomain Code=4 "“CoreTestCase” couldn’t be removed." UserInfo={NSUserStringVariant=(
        Remove
    ), NSFilePath=/Users/vagrant/Library/Developer/CoreSimulator/Devices/46C95008-E0C7-49E6-A4C1-8EFF7538A1C1/data/Containers/Data/Application/9969B8B3-097E-4E9A-8AD9-D28E45ED202A/tmp/CoreTestCase, NSUnderlyingError=0x6000211094a0 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}"
</details>
<details><summary>:x: XCTest failed: CoreTests/CourseSyncFilesInteractorLiveTests/testUpToDateFile</summary>


    failed: caught error: "Error Domain=NSCocoaErrorDomain Code=4 "“CoreTestCase” couldn’t be removed." UserInfo={NSUserStringVariant=(
        Remove
    ), NSFilePath=/Users/vagrant/Library/Developer/CoreSimulator/Devices/46C95008-E0C7-49E6-A4C1-8EFF7538A1C1/data/Containers/Data/Application/9969B8B3-097E-4E9A-8AD9-D28E45ED202A/tmp/CoreTestCase, NSUnderlyingError=0x600020c15470 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}"
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against e9086f9e5ed5b87515135a1fef8bb3837f2fcb86
</p>

**Review by szabinst — COMMENTED** [review]
I tested on a physical device and on simulator but didn't work on either of those. 

Observing the http requests, I see that `https://attilavarga-sandbox.instructuremedia.com/api/public/v1/courses/58625/media` returns with 200 but the download doesn't start afterwards. 

As a result there's only a blank area where the video should be. 

Simulator Screenshot - iPhone SE (3rd generation) - 2024-07-15 at 15 57 10

**Comment by vargaat** [comment]
> I tested on a physical device and on simulator but didn't work on either of those.
> 
> Observing the http requests, I see that `https://attilavarga-sandbox.instructuremedia.com/api/public/v1/courses/58625/media` returns with 200 but the download doesn't start afterwards.
> 
> As a result there's only a blank area where the video should be.
> 
 Fixed!

File: Core/Core/Todos/TodoListViewController.swift [hunk 1]
```diff
@@ -109,6 +110,7 @@ public class TodoListViewController: ScreenViewTrackableViewController, ErrorVie
         }
     }
 
+    private var subscriptions = Set<AnyCancellable>()
```
Comment (szabinst) [inline Core/Core/Todos/TodoListViewController.swift:113]: This is unused.

File: Core/Core/Todos/TodoListViewController.swift [hunk 2]
```diff
@@ -18,6 +18,7 @@
 
 import Foundation
 import UIKit
+import Combine
```
Comment (szabinst) [inline Core/Core/Todos/TodoListViewController.swift:21]: This is unused.

**Review by szabinst — COMMENTED** [review]
Nice work! The fix works now!
Shouldn't we follow our usual pattern with interactor protocols/implementations?

**Comment by vargaat** [comment]
> Shouldn't we follow our usual pattern with interactor protocols/implementations?

Refactored based on our discussion.

**Review by szabinst — APPROVED** [review]

**Review by ndrsszsz — APPROVED** [review]
QA +1
Tested on iPhone 11, iOS 17.5.1

**Pull request merged**